File: 开源控件框架库#@#/Retrofit网络框架/RxRetrofit/src/main/java/com/rxretrofit_build/net/RetrofitClient.java
package com.rxretrofit_build.net;

import android.content.Context;

import com.rxretrofit_build.factory.GsonConverterFactory;
import com.rxretrofit_build.model.BaseApiService;
import com.rxretrofit_build.model.DevelopBean;

import java.util.HashMap;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;


/**
 * Title:
 * Description:
 * Copyright:Copyright(c)2016
 * Company: Cree
 * CreateTime:16/11/28  09:41
 *
 * @author cree
 * @version 1.0
 */
public class RetrofitClient {

    private static Retrofit retrofit;
    private HashMap<String, Object> mapData;
    private static Context mContext;

    public static RetrofitClient getInstance(Context context) {
        mContext = context;
        return SingletonHolder.INSTANCE;
    }

    private static class SingletonHolder {
        private static RetrofitClient INSTANCE = new RetrofitClient(mContext);
    }


    public RetrofitClient() {
        this(null);
    }

    public RetrofitClient(Context context) {

        mapData = new HashMap<>();

        retrofit = new Retrofit
                .Builder()
                .client(genericClient())
                .baseUrl(BaseApiService.URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
//                .addConverterFactory(ScalarsConverterFactory.create())
                .build();
    }


    public OkHttpClient genericClient() {
        OkHttpClient.Builder interceptor = new OkHttpClient.Builder()
                .connectTimeout(5000, TimeUnit.MILLISECONDS)    //链接超时
                .writeTimeout(5000, TimeUnit.MILLISECONDS)       //写入超时
                .readTimeout(5000, TimeUnit.MILLISECONDS)       //读取超时
                //--------------------去InterceptorManager里面修改------！！！！！！！！！！！！1
                .addInterceptor(new InterceptorManager<HashMap<String, Object>>(mapData) {});
        if (DevelopBean.isDevelop) {
            interceptor.addInterceptor(new HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY));
        }

        return interceptor.build();
    }

    private BuilderReuqest builderReuqest;

    public void setBuilderReuqest(BuilderReuqest builderReuqest) {
        this.builderReuqest = builderReuqest;
    }

    public <T> T create(final Class<T> service) {
        if (service == null) {
            throw new RuntimeException("Api service is null!");
        }
        return retrofit.create(service);
    }

    public RetrofitClient setHashMap(HashMap<String, Object> hashMap) {
        mapData.clear();
        if (null == hashMap)
            return this;
        this.mapData.putAll(hashMap);
        return this;
    }
}

File: 自定义库/JNet/src/main/java/com/test/jasonchen/mysdk/net/client/RetrofitClient.kt
package com.test.jasonchen.mysdk.net.client

import com.test.jasonchen.mysdk.NetSDK
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

/**
 * Created by marti on 2018/3/4.
 */
object RetrofitClient {
    /*用来储存不同baseUrl的Retrofit实例*/
    private var retrofitMap: MutableMap<String, Retrofit> = mutableMapOf()
    /* baseUrlType 其实也就是 baseUrl*/
    var baseUrlType = NetSDK.baseUrl

    var retrofit: Retrofit = Retrofit
            .Builder()
            .client(OkClient.ok)                //设置Okhttp
            .baseUrl(baseUrlType)                          //设置baseUrl
            .addConverterFactory(GsonConverterFactory.create())     //支持GSON
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create()).build() //支持Rxjava
            /*
                每次获取的时候先判断之前是否保存过当前设置的baseUrlType的实例
                如果保存过就直接获取，没有保存过先重新设置baseUrlType，再保存
                最后把BaseUrlType依然改修为默认的，避免下次请求错用其他BaseURL
            */
        get() {
            if (retrofitMap[baseUrlType] == null) {
                field = field.newBuilder().baseUrl(baseUrlType).build()
                retrofitMap[baseUrlType] = field
            } else {
                field = retrofitMap[baseUrlType]!!
            }

            baseUrlType = NetSDK.baseUrl
            return field
        }
        private set


}